docs/egpu-switcher.1
.nh
.TH "egpu-switcher" "1" "Aug 2022" "egpu-switcher-0.0.1" ""

.SH NAME
.PP
egpu-switcher -


.SH SYNOPSIS
.PP
\fBegpu-switcher [flags]\fP


.SH DESCRIPTION

.SH OPTIONS
.PP
\fB-h\fP, \fB--help\fP[=false]
	help for egpu-switcher

.PP
\fB-v\fP, \fB--verbose\fP[=false]
	verbose output


.SH SEE ALSO
.PP
\fBegpu-switcher-cleanup(1)\fP, \fBegpu-switcher-config(1)\fP, \fBegpu-switcher-remove(1)\fP, \fBegpu-switcher-setup(1)\fP, \fBegpu-switcher-switch(1)\fP


.SH HISTORY
.PP
17-Aug-2022 Auto generated by spf13/cobra


debian/changelog
egpu-switcher (0.17.0) focal; urgency=medium

  * Fix some timing issues: if egpu is not connected on bootup, probe again every 500ms and give up after a few seconds instead of immediately (PR #52)
  * Fix "unbound variable" error when trying to switch to disconnected egpu (PR #53)
  * Remove previously installed egpu-switcher manpages on "make install" to prevent unnecessary user prompt (PR #54)
  * Fix issue in "if" statement when a user has customized their "Driver" part in xorg.conf.egpu (PR #57)

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 24 Jan 2021 16:38:19 +0100

egpu-switcher (0.16.0) focal; urgency=medium

  * Fix wrong egpu detection when a non-graphics device is attached to the registered PCI Bus ID

 -- Michael Hertig <info@michaelhertig.ch>  Tue, 06 Oct 2020 21:16:37 +0200

egpu-switcher (0.15.0) focal; urgency=medium

  * Update systemd service, hopefully fixing some timing issues on bootup
  * Add "egpu-switcher remove" method to remove eGPU without a reboot (experimental)
  * Update manpages

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 02 Aug 2020 14:15:15 +0200

egpu-switcher (0.14.0) focal; urgency=medium

  * No code changes: Just bumped the minor version to circumvent launchpad PPA publishing issue

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 14 Jun 2020 15:54:14 +0200

egpu-switcher (0.13.1) focal; urgency=medium

  * Fix bug in display detection which prevented to switch to the eGPU with AMD cards (thanks to ewagner12 for the fix)
  * Add "not recommended" hint to the "define a specific internal GPU" setup step

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 14 Jun 2020 13:53:25 +0200

egpu-switcher (0.13.0) disco; urgency=medium

  * Prevent error on cleanup if setup hasnt been executed
  * Remove unnecessary console print

 -- Michael Hertig <info@michaelhertig.ch>  Thu, 27 Feb 2020 23:33:13 +0100

egpu-switcher (0.12.0) disco; urgency=medium

  * Better GPU detection with "lspci -d" (thanks to xabolcs)
  * Prevent error on clean (thanks to maddocker)

 -- Michael Hertig <info@michaelhertig.ch>  Mon, 20 Jan 2020 20:24:59 +0100

egpu-switcher (0.11.1) disco; urgency=low

  * Previous build failed. Moved manpage to ./docs folder

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 09 Nov 2019 19:02:07 +0100

egpu-switcher (0.11.0) disco; urgency=low

  * Remove 'nvidia-xconfig' method of detecting GPUs
  * Change egpu.service to Type=oneshot
  * Make specifying the internal GPU optional (let X-Server figure it out themselves if EGPU is not connected)
  * Save chosen GPUs to config file to prevent re-run of setup after every update

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 09 Nov 2019 18:40:30 +0100

egpu-switcher (0.10.2) disco; urgency=medium

  * Fixed critical typo in Makefile which removed too many files

 -- Michael Hertig <info@michaelhertig.ch>  Mon, 26 Aug 2019 21:31:53 +0200

egpu-switcher (0.10.1) disco; urgency=medium

  * Needed to jump a version because of an invalid publish

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 23 Jun 2019 19:22:46 +0200

egpu-switcher (0.10.0) disco; urgency=medium

  * Fix minor bug in EGPU detection, add absolute path to systemd service 

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 23 Jun 2019 18:40:37 +0200

egpu-switcher (0.9.0) disco; urgency=medium

  * Fix bug in detecting GPUs with double-digit BUS-IDs, add 5s sleep before detection to give the egpu enough time to connect

 -- Michael Hertig <info@michaelhertig.ch>  Tue, 23 Apr 2019 21:12:50 +0200

egpu-switcher (0.8.0) disco; urgency=medium

  * Remove systemd service on cleanup

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 20 Apr 2019 12:16:04 +0200

egpu-switcher (0.6.0) disco; urgency=medium

  * Automatically choose the correct GPU driver

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 20 Apr 2019 11:47:09 +0200

egpu-switcher (0.5.0) disco; urgency=medium

  * Add systemd service, add lspci functionality as default

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 20 Apr 2019 02:56:08 +0200

egpu-switcher (0.4.0) bionic; urgency=medium

  * Add update-rc.d command to setup method to enable egpu-switcher on startup

 -- Michael Hertig <info@michaelhertig.ch>  Mon, 04 Mar 2019 20:29:34 +0100

egpu-switcher (0.3.0) bionic; urgency=medium

  * removed some lines from Makefile, added prerm script, extended cleanup method

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 17 Feb 2019 15:58:12 +0100

egpu-switcher (0.2.1) bionic; urgency=medium

  * Removed clean script from Makefile

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 17 Feb 2019 15:14:20 +0100

egpu-switcher (0.2.0) bionic; urgency=medium

  * Automatically install an init.d script to select the correct gpu on startup

 -- Michael Hertig <info@michaelhertig.ch>  Sun, 17 Feb 2019 14:57:09 +0100

egpu-switcher (0.1.0) bionic; urgency=medium

  * Initial Release.

 -- Michael Hertig <info@michaelhertig.ch>  Sat, 16 Feb 2019 21:21:09 +0100


egpu-switcher.sh
#!/usr/bin/env bash

set -o errexit # exit script if a command fails
set -o nounset # exit when trying to use undeclared variables
#set -o xtrace # debug

# define some colors
declare red='\033[1;31m'
declare yellow='\033[1;33m'
declare green='\033[1;32m'
declare blue='\033[1;34m'
declare blank='\033[0m'

# define log level prefix
declare error="$red[error]$blank"
declare warn="$yellow[warn]$blank"
declare success="$green[success]$blank"
declare info="$blue[info]$blank"

# misc
declare datetime=$(date '+%Y%m%d%H%M%S')
declare number_regex='^[0-9]+$'

# constant variables for X-Server files
declare xdir=/etc/X11
declare xfile=$xdir/xorg.conf
declare xfile_egpu=$xdir/xorg.conf.egpu
declare xfile_internal=$xdir/xorg.conf.internal
declare xfile_backup=$xfile.backup

# constant variables for temp files
declare tmp_template=/usr/share/egpu-switcher/xorg.conf.template

# constant variables for systemd files
declare systemd_template=/usr/share/egpu-switcher/egpu.service
declare systemd_folder=/etc/systemd/system

# todo: this assumes that there is bash 4+ installed
declare -A gpus=()
declare gpu_connected=0

# helper method for printing error messages
function print_error() {
	echo -e "$error $1"
}

# helper method for printing info messages
function print_info() {
	echo -e "$info $1"
}

# helper method for printing warn messages
function print_warn() {
	echo -e "$warn $1"
}

# helper method for printin success messages
function print_success() {
	echo -e "$success $1"
}

# config
config_dir=/etc/egpu-switcher
config_file=${config_dir}/egpu-switcher.conf
typeset -A config
config=(
	[internal_gpu]=""
	[internal_driver]=""
	[external_gpu]=""
	[external_driver]=""
)

# read from config file
if [ -f $config_file ]; then
	#print_info "Reading contents of your current configuration file at '${config_file}'"
	while read line
	do
		if echo $line | grep -F = &>/dev/null
		then
			varname=$(echo "$line" | cut -d '=' -f 1)
			config[$varname]=$(echo "$line" | cut -d '=' -f 2-)
		fi
	done < $config_file
fi

# check if the script is run as root
if [[ $EUID -ne 0 ]]; then 
  	print_error "You need to run the script with root privileges"
  	exit
fi

# check if an argument was passed
if [ -z ${1+x} ]; then
	print_error "No argument passed."
	exit
fi

# read the connected GPUs and put them into the "gpus" associative array
function read_gpus() {

	# empty the gpus array
	gpus=()

	declare lines=$(lspci -D -d ::0300 && lspci -D -d ::0302)
	while read -r line ; do
		declare name=$(echo $line | grep -o -e "[^:]*$")
		declare bus=$(echo $line | grep -o -e "^[^ ]*")

		# The bus IDs in hex
		declare bus1h=${bus:5:2}
		declare bus2h=${bus:8:2}
		declare bus3h=${bus:11:1}

		# The bus IDs in dec
		declare bus1d=$((16#$bus1h))
		declare bus2d=$((16#$bus2h))
		declare bus3d=$((16#$bus3h))

		# Remove the whitespace at the beginning of the name
		# And concatenate bus IDs
		bus="${bus1d}:${bus2d}:${bus3d}"
		name=${name:1}

		# Put the result into the gpus array
		gpus+=( [$bus]=$name )
	done <<< $lines
	return
}

# returns 1 if egpu is connected, 0 if not
function is_egpu_connected() {

	# read pci id from xorg.conf.egpu
	declare egpu_pci_id=$(cat $xfile_egpu | grep -Ei "BusID" | grep -oEi '[0-9]+\:[0-9]+\:[0-9]+')

	# create an array by splitting the BUS-ID on ':'
	declare busArray=(${egpu_pci_id//:/ })
	declare bus1d=${busArray[0]}
	declare bus2d=${busArray[1]}
	declare bus3d=${busArray[2]}

	# convert dec to hex
	declare bus1h=$(printf "%02x" $bus1d)
	declare bus2h=$(printf "%02x" $bus2d)
	declare bus3h=$(printf "%01x" $bus3d)

	# instantiate counter
	declare i=1

	# begin infinite loop to allow retries if the egpu isn't connected immediately on bootup
	while [ true ]; do

		# if a video device is connected to the BUS-ID
		if [ $( (lspci -d ::0300 && lspci -d ::0302) | grep -iEc "$bus1h:$bus2h.$bus3h") -eq 1 ]; then
			print_info "EGPU is ${green}connected${blank}."
			gpu_connected=1
			hex_id=$bus1h:$bus2h.$bus3h
			break
		else
			# escape the infinite loop after a certain amount of retries
			if [ $i -ge 6 ]; then
				print_info "EGPU is ${red}disconnected${blank}."
				gpu_connected=0
				hex_id=$bus1h:$bus2h.$bus3h
				break
			fi
		fi

		# increase counter by 1
		i=$(( $i + 1 ))

		# sleep for 500ms before retrying
		sleep 0.5
	done
}

# get the matching driver according to the gpu name
function get_driver() {

	input=${1}

	if [ $(echo "${input}" | grep -Eic "nvidia") -gt 0 ]; then
		echo "nvidia"
		return
	fi

	if [ $(echo "${input}" | grep -Eic "intel") -gt 0 ]; then
		echo "intel"
		return
	fi

	if [ $(echo "${input}" | grep -Eic "amd") -gt 0 ]; then
		echo "amdgpu"
		return
	fi
}

# prompts the user to define their external/internal GPUs
# and saves it into the config file
function configure() {

	# reset current config
	config[internal_gpu]=""
	config[internal_driver]=""
	config[external_gpu]=""
	config[external_driver]=""

	# read currently attached GPUs
	read_gpus

	# save the number of lines to a variable
	declare num_of_gpus=${#gpus[@]}

	# additional check
	if [ $num_of_gpus -lt "2" ]; then
		print_warn "Only ${num_of_gpus} GPUs found, there need to be at least 2. Make sure to connect your EGPU for the setup."
		exit
	fi

	# print the GPUs found
	echo ""
	echo -e "Found $num_of_gpus possible GPUs..."
	echo ""

	declare mapping=()
	declare i=0
	for key in ${!gpus[@]}; do
		i=$((i+1))
		mapping+=([${i}]=${key})
		echo "  $i: ${gpus[${key}]} (${key})"
	done

	echo ""

	printf "Would you like to define a specific$green INTERNAL$blank GPU? (not recommended) [y/N]: "
	read specify_internal
	if [[ $specify_internal == "y" ]]; then
		# prompt to choose the internal gpu from the list
		printf "Choose your preferred$green INTERNAL$blank GPU [1-$num_of_gpus]: "
		read internal
		declare full_internal=${gpus[${mapping[$internal]}]}
		declare pci_internal=${mapping[$internal]}

		if ! [[ $internal =~ $number_regex ]] || [ -z "$pci_internal" ]; then
			print_error "Your input is invalid. Exiting setup..."
			exit
		fi	

		config[internal_gpu]=${pci_internal}
		config[internal_driver]=$(get_driver "$full_internal")

		if [ -z ${config[internal_driver]} ]; then
			print_info "Could not parse manufacturer from \"$full_internal\"."
			printf "Please manually enter the driver to be used: "
			read driver
			config[internal_driver]=${driver}
		fi
	fi

	# prompt to choose the external gpu from the list
	printf "Choose your preferred$green EXTERNAL$blank GPU [1-$num_of_gpus]: "
	read external
	declare full_external=${gpus[${mapping[$external]}]}
	declare pci_external=${mapping[$external]}

	if ! [[ $external =~ $number_regex ]] || [ -z "$pci_external" ]; then
		print_error "Your input is invalid. Exiting setup..."
		exit
	fi

	config[external_gpu]=${pci_external}
	config[external_driver]=$(get_driver "$full_external")

	if [ -z "${config[external_driver]}" ]; then
		print_info "Could not parse manufacturer from \"$full_external\"."
		printf "Please manually enter the driver to be used: "
		read driver
		config[external_driver]=${driver}
	fi

	echo ""

	# create config directory if it doesnt exist
	mkdir -p $config_dir

	# empty current config file
	true > $config_file

	# write new configurations to config file
	for key in ${!config[@]}; do
		echo "${key}=${config[${key}]}" >> $config_file
	done

	print_info "Saved new configuration to ${config_file}"
}

function setup() {

	declare override=${1}
	declare noprompt=${2}

	# check if the template/script files can be found
	if [ ! -f $tmp_template ]; then
		print_error "The file $tmp_template wasn't found."
		exit
	fi

	if [ -z ${config[external_gpu]} ]; then
		if [ ${noprompt} -eq 1 ]; then
			print_warn "It seems like you haven't configured egpu-switcher yet. Please run 'egpu-switcher setup' first."
			exit
		else
			configure
		fi
	else
		print_info "Using existing configuration file at '${config_file}''."
		print_info "If you want to reconfigure egpu-switcher, please run 'egpu-switcher config'."
	fi

	# create the internal xorg config file
	if [ ! -z ${config[internal_gpu]} ]; then
		cp $tmp_template $xfile_internal
		sed -i -e 's/\$BUS/'${config[internal_gpu]}'/g' -e 's/\$DRIVER/'${config[internal_driver]}'/g' -e 's/\$ID/Device0/g' $xfile_internal
	else 
		true > $xfile_internal
	fi

	# create the external xorg config file
	cp $tmp_template $xfile_egpu
	sed -i -e 's/\$BUS/'${config[external_gpu]}'/g' -e 's/\$DRIVER/'${config[external_driver]}'/g' -e 's/\$ID/Device0/g' $xfile_egpu

	# Executing the switch command to create the xorg.conf file
	switch auto ${override}

	# setup startup script
	cp $systemd_template $systemd_folder
	systemctl daemon-reload
	systemctl enable egpu.service

	print_success "Done... Setup finished"
}

function switch() {

	declare mode=${1}
	declare override=${2}
	
	# Check if the xorg.conf files for internal and egpu exist
	if ! [ -f $xfile_egpu ] || ! [ -f $xfile_internal ]; then
		print_error "The xorg.conf files for egpu and internal do not exist. Run the setup first."
		return
	fi

	# Check if there is a xorg.conf file, and back it up
	if [ -f $xfile ] && ! [ -L $xfile ]; then
		print_warn "The $xfile file already exists. Saving a backup to $xfile_backup.$datetime"
		cp "$xfile" "$xfile_backup.$datetime"
	fi

	# if no parameter was passed to the method
	if [ ${mode} = "auto" ]; then

		print_info "Automatically detecting if egpu is connected... "
		is_egpu_connected
		if [ ${gpu_connected} = 1 ] ; then
			mode="egpu"
		else
			mode="internal"
		fi
	fi

	# when mode is 'egpu' and no 'nvidia' driver is used
	if [ ${mode} = "egpu" ] && ! grep -Eiq 'Driver.*nvidia' $xfile_egpu; then

		if [ -z ${hex_id+x} ]; then
			is_egpu_connected
		fi
		
		declare disp_path=/sys/bus/pci/devices/[0-9a-f:]*${hex_id}/drm/card[0-9]*/card[0-9]*-*/status
		declare disp_num=0
		declare disp_disconnect=0
		for disp in ${disp_path}; do
			if [ -e $disp ]; then
				((disp_num = $disp_num + 1)) || true
				((disp_disconnect = $disp_disconnect + $(cat $disp | grep -ce ^disconnected$))) || true
			fi
		done

		if [ $disp_disconnect -eq $disp_num ] && [ $disp_num -gt 0 ]; then
			print_warn "No eGPU attached display detected with open source drivers. (Of ${disp_num} eGPU outputs detected) Internal mode and setting DRI_PRIME variable are recommended for this configuration."
			if [ ${override} -eq 1 ]; then
				print_warn "-> Overridden: Setting eGPU mode"
				mode="egpu"
			else 
				print_warn "Run 'egpu-switcher switch egpu --override' to force loading eGPU mode"
				print_warn "-> Not setting eGPU mode."
				mode="internal"
			fi
		fi
	fi

	if [ ${mode} = "egpu" ]; then
		print_info "Create symlink ${green}${xfile}${blank} -> ${xfile_egpu}"
		ln -sf ${xfile_egpu} ${xfile}
		return
	fi

	if [ ${mode} = "internal" ]; then
		print_info "Create symlink ${green}${xfile}${blank} -> ${xfile_internal}"
		ln -sf ${xfile_internal} ${xfile}
		return
	fi

	print_error "The argument '${mode}' that was passed to the switch() method is not valid."
}

function remove() {

	if [ -z ${hex_id+x} ]; then
		is_egpu_connected
	fi

	if [ $gpu_connected = 0 ]; then
		print_error "No eGPU detected at BusID specified in xorg.conf.egpu. Stopping removal."
		exit
	fi

	# Find GPU and audio devices
	declare device_id=$(echo ${hex_id} | cut -f 1 -d '.').
	declare device_path=/sys/bus/pci/devices/[0-9a-f:]*${device_id}[0-9]*/remove
	declare vga_driver=$(cat $xfile_egpu | grep -Ei "Driver" | cut -f 2 -d \")

	# Do actual GPU removal
	( trap '' HUP TERM
		while [ "$(systemctl status display-manager | awk '/Active:/{print$2}')" \
			= "active" ]; do
			sleep 1
		done

		if [ ${vga_driver} = "nvidia" ]; then
			systemctl stop nvidia-persistenced.service
			if [ $(lsmod | grep "nvidia_uvm " | awk '{print $3}') -gt 0 ] || [ $(lsmod | grep "nvidia_drm " | awk '{print $3}') -gt 0 ]; then
				systemctl start display-manager.service
				print_error "Driver still in use. Check for applications like Folding@Home running in the background. Stopping removal."
				exit
			fi
			for drivers in nvidia_uvm nvidia_drm nvidia_modeset nvidia; do
				modprobe -r ${drivers}
			done
		else
			if [ $(lsmod | grep "${vga_driver} " | awk '{print $3}') -gt 0 ]; then
				systemctl start display-manager.service
				print_error "Driver still in use. Check for applications like Folding@Home running in the background. Stopping removal."
				exit
			fi
			modprobe -r ${vga_driver}
		fi

		for dev_paths in ${device_path}; do
			if [ -e $dev_paths ]; then
				echo 1 > $dev_paths
			fi
		done

		if [ $(lspci -k | grep -c ${vga_driver}) -gt 0 ]; then
			modprobe ${vga_driver}
			if [ ${vga_driver} = "nvidia" ]; then
				modprobe nvidia_drm
			fi
			sleep 1
		fi

		systemctl start display-manager.service ) &
	systemctl stop display-manager.service
}

function cleanup() {

	declare hard=${1}

	print_info "Starting cleanup process"
	rm -f ${xfile_egpu}
	rm -f ${xfile_internal}

	# delete the xorg.conf file, if it is a symlink and restore the last backup
    if [ -L ${xfile} ]; then
        rm -f ${xfile}
        if [ -e ${xfile_backup}.* ]; then
            local lastbackup=$(ls -t ${xfile_backup}.* | head -1)
            print_info "Restoring latest backup ${green}${lastbackup}"
            mv ${lastbackup} ${xfile}
        fi
    fi

	if [ ${hard} -eq 1 ]; then
		print_info "Removing configuration files (--hard)"
		rm -f ${config_file}
		rm -fd ${config_dir}
	fi

	# only try to stop the egpu.service if its loaded. 
	# note: using sed rather than the --value property on purpose, to support older versions of systemd
	if [ $(sudo systemctl show -p LoadState egpu.service | sed 's/LoadState=//g') == "loaded" ]; then
		print_info "Removing the 'egpu.service' systemd service"
		systemctl stop egpu.service
		systemctl disable egpu.service
		rm ${systemd_folder}/egpu.service
		systemctl daemon-reload
		systemctl reset-failed
	fi

	print_success "Done... Finished cleanup"
}


if [ $1 = "setup" ]; then

	declare override=0
	declare noprompt=0
	if [ $# -gt 1 ]; then
		for option in "$@"
		do
			if [ ${option} = "--override" ]; then
				override=1
			elif [ ${option} = "--noprompt" ]; then
				noprompt=1
			fi
		done
	fi
	setup ${override} ${noprompt}

elif [ $1 = "switch" ]; then

	declare override=0
	if [ $# -lt 2 ]; then
		print_error "No argument passed to the switch method. Possible options: auto, egpu, internal"
		exit
	else
		for option in "$@" 
		do
			if [ ${option} = "--override" ]; then
				override=1
			fi
		done
	fi
	switch ${2} ${override}

elif [ $1 = "config" ]; then

	configure

elif [ $1 = "cleanup" ]; then

	declare hard=0
	if [ $# -gt 1 ]; then 
		for option in "$@"
		do
			if [ ${option} = "--hard" ]; then
				hard=1
			fi
		done
	fi

	cleanup ${hard}

elif [ $1 = "remove" ]; then

	print_warn "This will switch to internal mode, remove the eGPU PCIe addresses and log out all users. Continue? [y/N]: "
	read specify_remove
	if [[ $specify_remove == "y" ]]; then
		switch internal 0
		remove
	fi

else
	print_error "Unknown argument '$1'.\navailable commands: setup, switch, config, cleanup, remove"
fi

# systemctl restart display-manager.service
